File: Code/Chapter-13/envmapping.cpp
//
// Example of environment mapping
//
// Author: Alexey Boreskov <steps3d@gmail.com>, <steps3d@narod.ru>
//

#include	"GlutRotateWindow.h"
#include	"Program.h"
#include	"BasicMesh.h"
#include	"Texture.h"

class	MeshWindow : public GlutRotateWindow
{
	Program     program;
	BasicMesh * mesh = nullptr;
	Texture     tex;
	
public:
	MeshWindow () : GlutRotateWindow ( 200, 50, 900, 900, "Env. mapped torus" )
	{
		std::string	texName = "../../Textures/Cubemaps/Snow.dds";
	
		if ( !tex.loadCubemap ( texName ) )
			exit ( "Error loading texture %s\n", texName.c_str () );
		
		if ( !program.loadProgram ( "envmapping.glsl" ) )
			exit ( "Error building program: %s\n", program.getLog ().c_str () );
		
		program.bind       ();
		program.setTexture ( "image", 0 );
		
//		mesh = createTorus ( 2, 4, 30, 30  );
		mesh  = createKnot ( 1, 4, 120, 30 );
		eye  = glm::vec3 ( 5, 5, 5 );
	}

	void redisplay () override
	{
		glClear ( GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT );

		glm::mat4	mv = getRotation  ();
		glm::mat3	nm = normalMatrix ( mv );
	
		tex.bind                 ();
		program.setUniformMatrix ( "mv",  mv );
		program.setUniformMatrix ( "nm",  nm );
		program.setUniformVector ( "eye", eye );

		mesh -> render ();
		
		tex.unbind     ();
	}

	void reshape ( int w, int h ) override
	{
		GlutWindow::reshape ( w, h );
		
		program.setUniformMatrix ( "proj", getProjection () );
	}
};

int main ( int argc, char * argv [] )
{
	GlutWindow::init ( argc, argv );
	
	MeshWindow	win;
	
	return GlutWindow::run ();
}

File: Code/python-code/envmapping.py
import math
import glm
import numpy
from OpenGL.GL import *
import Window
import Program
import Texture
import Mesh
import dds

class   MyWindow ( Window.RotationWindow ):
    def __init__ ( self, w, h, t ):
        super().__init__ ( w, h, t )
        self.mesh      = Mesh.Mesh.createKnot ( 1, 4, 120, 30 )
        self.texture, target, w, h, d = dds.readDdsFile ( '../../Textures/Cubemaps/LobbyCube.dds' )
        self.shader    = Program.Program ( glsl = "envmapping.glsl" )
        glBindTexture ( GL_TEXTURE_CUBE_MAP, self.texture )
        self.shader.use         ()
        self.shader.setTexture  ( "image", 0 )

    def redisplay ( self ):
        glClearColor ( 0.2, 0.3, 0.2, 1.0 )
        glClear      ( GL_COLOR_BUFFER_BIT + GL_DEPTH_BUFFER_BIT )
        glEnable     ( GL_DEPTH_TEST )

        self.shader.setUniformMat ( "mv",  self.getRotation () )
        self.shader.setUniformMat ( "nm",  self.normalMatrix ( self.getRotation () ) )
        self.shader.setUniformVec ( "eye", self.eye )
        self.mesh.render()

    def reshape ( self, width, height ):
        super().reshape ( width, height )
        self.shader.setUniformMat ( "proj",  self.getProjection () )

    def mouseScroll ( self, dx, dy ):
        self.eye += glm.vec3 ( 0.1 * ( 1 if dy >= 0 else -1 ) )
        self.reshape ( self.width, self.height )

def main():
    win = MyWindow ( 900, 900, "Environmapped knot" )
    win.run ()

if __name__ == "__main__":
    main()

